src/typechecker/filter_map.rs

```diff
@@ -0,0 +1,460 @@
+                };
+                scope.insert_var(field_name, ty)?;
+            }
+            ast::RxTxType::Split(rx, tx) => {
+                let Some(ty) = self.get_type(&rx.ty) else {
+                    return Err("type for rx is not defined".into());
+                };
+                scope.insert_var(&rx.field_name, ty)?;
+
+                let Some(ty) = self.get_type(&tx.ty) else {
+                    return Err("type for tx is not defined".into());
+                };
+                scope.insert_var(&tx.field_name, ty)?;
+            }
+            ast::RxTxType::PassThrough(ast::TypeIdentField {
+                field_name,
+                ty,
+            }) => {
+                let Some(ty) = self.get_type(ty) else {
+                    return Err("type for rx_tx is not defined".into());
+                };
+                scope.insert_var(field_name, ty)?;
+            }
+        }
+
+        for (ident, expr) in assignments {
+            let t = self.expr(&scope, expr)?;
+            scope.insert_var(ident, t)?;
+        }
+
+        Ok(())
+    }
+
+    fn term_section(
+        &mut self,
+        scope: &Scope,
+        term_section: &ast::TermSection,
+    ) -> TypeResult<(String, Type)> {
+        let ast::TermSection {
+            ident,
+            for_kv: _,
+            with_kv,
+            body: ast::TermBody { scopes },
+        } = term_section;
+
+        let mut scope = scope.wrap();
+
+        let args = self.with_clause(&mut scope, &with_kv)?;
+
+        for ast::TermScope {
+            scope: _,
+            operator,
+            match_arms,
+        } in scopes
+        {
+            match operator {
+                ast::MatchOperator::Match => {
+                    for (pattern, exprs) in match_arms {
+                        assert!(pattern.is_none(), "ICE");
+                        for expr in exprs {
+                            // We ignore the result because it
+                            // must be boolean.
+                            self.logical_expr(&scope, expr)?;
+                        }
+                    }
+                }
+                ast::MatchOperator::MatchValueWith(ast::Identifier {
+                    ident,
+                }) => {
+                    let x = scope.get_var(ident)?;
+                    let Type::Enum(_, variants) = self.resolve_type(x) else {
+                        return Err(format!("Cannot match on the type '{x:?}', because only matching on enums is supported."));
+                    };
+
+                    // We'll keep track of used variants to do some basic
+                    // exhaustiveness checking. Only a warning for now.
+                    let mut used_variants = Vec::<&str>::new();
+
+                    for (pattern, exprs) in match_arms {
+                        let ast::TermPatternMatchArm {
+                            variant_id,
+                            data_field,
```

Comment:
Maybe you want to get rid of the unwrap(), or add a comment why it's infallible?